core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -672,24 +681,42 @@ public String toString()
         // This should not be a ConcurrentHashMap because we want to prevent concurrent addition of new consumers during the
         // removal of existing consumers from this map in addDynamicFilters. This ensures that new consumers don't miss filter completion.
         private final Map<DynamicFilterId, List<Consumer<Map<DynamicFilterId, Domain>>>> dynamicFilterConsumers = new HashMap<>();
-        private final long queryStartTime = System.nanoTime();
+        private long queryAttemptStartTime = System.nanoTime();
 
         private DynamicFilterContext(
                 Session session,
                 Set<DynamicFilterId> dynamicFilters,
-                Map<DynamicFilterId, SettableFuture<Void>> lazyDynamicFilters,
+                Set<DynamicFilterId> lazyDynamicFilters,
                 Set<DynamicFilterId> replicatedDynamicFilters)
         {
             this.session = requireNonNull(session, "session is null");
             this.dynamicFilters = requireNonNull(dynamicFilters, "dynamicFilters is null");
-            this.lazyDynamicFilters = requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null");
+            this.lazyDynamicFilters = requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null").stream()
+                .collect(toMap(identity(), filter -> SettableFuture.create()));
             this.replicatedDynamicFilters = requireNonNull(replicatedDynamicFilters, "replicatedDynamicFilters is null");
             dynamicFilters.forEach(filter -> {
                 taskDynamicFilters.put(filter, new ConcurrentHashMap<>());
                 dynamicFilterConsumers.put(filter, new ArrayList<>());
             });
         }
 
+        void resetForQueryRetry()
+        {
+            dynamicFilterSummaries.clear();
+            dynamicFilterCollectionTime.clear();
+            lazyDynamicFilters.replaceAll((filterId, currentFuture) -> {
+                currentFuture.set(null);
```

Comment:
Is it necessary? Setting a future would trigger listeners to be processed. And that would happen under the lock (from `createDistributedStagesScheduler`) that is usually not recommended? Is it okay to avoid unblocking the future? The previous attempt (including all the tasks and the split enumeration will be closed and terminated anyway)

Reply:
It's not necessary, got rid of it now